File: ga/glm4_result_2024_6/result/queue_glm/11_queue.c
use std::collections::VecDeque;

type ElementType = i32;

const MIN_QUEUE_SIZE: usize = 5;

struct Queue {
    elements: VecDeque<ElementType>,
}

impl Queue {
    fn new(max_elements: usize) -> Queue {
        if max_elements < MIN_QUEUE_SIZE {
            panic!("Queue size is too small");
        }
        Queue {
            elements: VecDeque::with_capacity(max_elements),
        }
    }

    fn is_empty(&self) -> bool {
        self.elements.is_empty()
    }

    fn is_full(&self) -> bool {
        self.elements.len() == self.elements.capacity()
    }

    fn enqueue(&mut self, element: ElementType) {
        if self.is_full() {
            panic!("Full queue");
        }
        self.elements.push_back(element);
    }

    fn front(&self) -> ElementType {
        if self.is_empty() {
            panic!("Empty queue");
        }
        self.elements.front().unwrap()
    }

    fn dequeue(&mut self) {
        if self.is_empty() {
            panic!("Empty queue");
        }
        self.elements.pop_front();
    }

    fn front_and_dequeue(&mut self) -> ElementType {
        if self.is_empty() {
            panic!("Empty queue");
        }
        self.elements.pop_front().unwrap()
    }
}

fn main() {
    let mut q = Queue::new(12);

    for i in 0..10 {
        q.enqueue(i);
    }

    while !q.is_empty() {
        println!("{}", q.front());
        q.dequeue();
    }

    for i in 0..10 {
        q.enqueue(i);
    }

    while !q.is_empty() {
        println!("{}", q.front());
        q.dequeue();
    }
}

File: ga/4omini_result_2024_7/result/queue_mini/11_queue.rs
use std::io::{self, Write};

#[derive(Debug)]
struct Queue {
    capacity: usize,
    front: usize,
    rear: usize,
    size: usize,
    array: Vec<i32>, // Using Vec for dynamic array
}

impl Queue {
    fn new(max_elements: usize) -> Result<Queue, String> {
        if max_elements < 5 {
            return Err("Queue size is too small".to_string());
        }
        Ok(Queue {
            capacity: max_elements,
            front: 1, // We'll use 1-based indexing for consistency with the C code
            rear: 0,
            size: 0,
            array: vec![0; max_elements],
        })
    }

    fn is_empty(&self) -> bool {
        self.size == 0
    }

    fn is_full(&self) -> bool {
        self.size == self.capacity
    }

    fn make_empty(&mut self) {
        self.size = 0;
        self.front = 1;
        self.rear = 0;
    }

    fn enqueue(&mut self, x: i32) -> Result<(), String> {
        if self.is_full() {
            return Err("Full queue".to_string());
        }
        self.size += 1;
        self.rear = self.succ(self.rear);
        self.array[self.rear] = x;
        Ok(())
    }

    fn front(&self) -> Result<i32, String> {
        if self.is_empty() {
            return Err("Empty queue".to_string());
        }
        Ok(self.array[self.front])
    }

    fn dequeue(&mut self) -> Result<(), String> {
        if self.is_empty() {
            return Err("Empty queue".to_string());
        }
        self.size -= 1;
        self.front = self.succ(self.front);
        Ok(())
    }

    fn front_and_dequeue(&mut self) -> Result<i32, String> {
        if self.is_empty() {
            return Err("Empty queue".to_string());
        }
        let x = self.array[self.front];
        self.size -= 1;
        self.front = self.succ(self.front);
        Ok(x)
    }

    fn succ(&self, value: usize) -> usize {
        if value + 1 == self.capacity {
            0
        } else {
            value + 1
        }
    }
}

fn main() {
    let mut q = Queue::new(12).expect("Failed to create queue");

    for i in 0..10 {
        q.enqueue(i).expect("Failed to enqueue");
    }

    while !q.is_empty() {
        let front_value = q.front().expect("Failed to get front");
        println!("{}", front_value);
        q.dequeue().expect("Failed to dequeue");
    }

    for i in 0..10 {
        q.enqueue(i).expect("Failed to enqueue");
    }

    while !q.is_empty() {
        let front_value = q.front().expect("Failed to get front");
        println!("{}", front_value);
        q.dequeue().expect("Failed to dequeue");
    }
}